guides/keras_cv/object_detection_keras_cv.py

```diff
@@ -0,0 +1,872 @@
+
+"""
+### Loss functions
+
+You may not be familiar with the `"focal"` or `"smoothl1"` losses.  While not
+common in other models, these losses are more or less staples in the object
+detection world.
+
+In short, ["Focal Loss"](https://arxiv.org/abs/1708.02002) places extra emphasis
+on difficult training examples.  This is useful when training the classification
+loss, as the majority of the losses are assigned to the background class.
+
+"SmoothL1 Loss" is used to [prevent exploding gradients](https://arxiv.org/abs/1504.08083)
+that often occur when attempting to perform the box regression task.
+
+In KerasCV you can use these losses simply by passing the strings `"focal"` and
+`"smoothl1"` to `compile()`:
+"""
+
+pretrained_model.compile(
+    classification_loss="focal",
+    box_loss="smoothl1",
+)
+
+"""
+### Metric evaluation
+
+Just like any other metric, you can pass the `KerasCV` object detection metrics
+to `compile()`.  The most popular Object Detection metrics are COCO metrics,
+which were published alongside the MSCOCO dataset.  KerasCV provides an easy
+to use suite of COCO metrics. under the `keras_cv.metrics.BoxCOCOMetrics`
+symbol:
+"""
+
+coco_metrics = keras_cv.metrics.BoxCOCOMetrics(
+    bounding_box_format="xywh", evaluate_freq=20
+)
+
+"""
+Let's define a quick helper to print our metrics in a nice table:
+"""
+
+
+def print_metrics(metrics):
+    maxlen = max([len(key) for key in result.keys()])
+    print("Metrics:")
+    print("-" * (maxlen + 1))
+    for k, v in metrics.items():
+        print(f"{k.ljust(maxlen+1)}: {v.numpy():0.2f}")
+
+
+"""
+Due to the high computational cost of computing COCO metrics, the KerasCV
+`BoxCOCOMetrics` component requires an `evaluate_freq` parameter to be passed to
+its constructor.  Every `evaluate_freq`-th call to `update_state()`, the metric
+will recompute the result.  In between invocations, a cached version of the
+result will be returned.
+
+To force an evaluation, you may call `coco_metrics.result(force=True)`:
+"""
+
+pretrained_model.compile(
+    classification_loss="focal",
+    box_loss="smoothl1",
+    optimizer=optimizer,
+    metrics=[coco_metrics],
+)
+coco_metrics.reset_state()
+result = pretrained_model.evaluate(eval_ds.take(40), verbose=0)
+result = coco_metrics.result(force=True)
+
+print_metrics(result)
+
+"""
+
+**A note on TPU compatibility:**
+
+Evaluation of `BoxCOCOMetrics` require running `tf.image.non_max_suppression()`
+inside of the `model.train_step()` and `model.evaluation_step()` functions.
+Due to this, the metric suite is not compatible with TPU when used with the
+`compile()` API.
+
+Luckily, there are two workarounds that allow you to still train a RetinaNet on TPU:
+
+- The use of a custom callback
+- Using a [SideCarEvaluator](https://www.tensorflow.org/api_docs/python/tf/keras/utils/SidecarEvaluator)
+
+Let's use a custom callback to achieve TPU compatibility in this guide:
+"""
+
+
+class EvaluateCOCOMetricsCallback(keras.callbacks.Callback):
```

Comment:
This is pretty bad. It means that all KerasCV OD examples will be TPU-incompatible out of the box.